### ISSUE TITLE
Type infomation is missed when recursively use generic `type & interface`

### ISSUE THREAD

**Issue by xcatliu** [Microsoft/TypeScript#23706]
**TypeScript Version:**  2.8.3

**Search Terms:** type keyof extra 

**Code**

```ts
type Template<T> = {
    [K in keyof T]: Template<T[K]>;
} & {
    __name?: string;
}

const template: Template<{
    foo: number;
    bar: {
        baz: string;
    }
}> = {
    foo: 1,
    __name: 'n',
    bar: {
        baz: 'b',
        not_exist_key: true
    }
};
```

**Expected behavior:** Should report error: `'not_exist_key' does not exist in type 'Template<{ baz: string; }>'`

**Actual behavior:** No errors

**Playground Link:**  [playground](https://www.typescriptlang.org/play/#src=type%20Template%3CT%3E%20%3D%20%7B%0D%0A%20%20%20%20%5BK%20in%20keyof%20T%5D%3A%20Template%3CT%5BK%5D%3E%3B%0D%0A%7D%20%26%20%7B%0D%0A%20%20%20%20__name%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Aconst%20template%3A%20Template%3C%7B%0D%0A%20%20%20%20foo%3A%20number%3B%0D%0A%20%20%20%20bar%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20baz%3A%20string%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%3E%20%3D%20%7B%0D%0A%20%20%20%20foo%3A%201%2C%0D%0A%20%20%20%20__name%3A%20'n'%2C%0D%0A%20%20%20%20bar%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20baz%3A%20'b'%2C%0D%0A%20%20%20%20%20%20%20%20not_exist_key%3A%20true%0D%0A%20%20%20%20%7D%0D%0A%7D%3B)

**Related Issues:**

**Comment by xcatliu** [comment]
@jack-williams Thanks for your reply. The smaller example make this issue clearer.
Another question is why can't I add an extra key to mapped types like this:

```ts
type Template<T> = {
    [K in keyof T]: Template<T[K]>;
    foo?: string;
}
// Cannot find name 'foo'.
// '}' expected.
```

**Comment by mhegazy** [comment]
Duplicate of #13813. thanks @jack-williams!
